File: 124. Binary Tree Maximum Path Sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        total, path = self.getTotal(root)
        return total
    
    def getTotal(self, root):
        if root.left and root.right:
            lsum, lpath = self.getTotal(root.left)
            rsum, rpath = self.getTotal(root.right)
            
            mypath = max(lpath + root.val, rpath + root.val, root.val)
            mysum = max(mypath, rsum, lsum, root.val + lpath + rpath)
            
            return mysum, mypath
        elif not root.left and root.right:
            rsum, rpath = self.getTotal(root.right)
            
            mypath = max(rpath + root.val, root.val)
            mysum = max(mypath, rsum)
            return mysum, mypath
        elif root.left and not root.right:
            lsum, lpath = self.getTotal(root.left)
            
            mypath = max(lpath + root.val, root.val)
            mysum = max(mypath, lsum)
            
            return mysum, mypath
        else:
            return root.val, root.val

File: 124. Binary Tree Maximum Path Sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        } else {
            Result result = recurse(root);
            return result.maxSum;
        }
    }
    
    public Result recurse(TreeNode root) {
        Result result = new Result(root.val, root.val);
        
        if (root.left == null && root.right == null) {
            ;
        } else if (root.left != null && root.right == null) {
            Result left = recurse(root.left);
            result.maxPath = Math.max(root.val, root.val + left.maxPath);
            result.maxSum = Math.max(result.maxPath, left.maxSum);
        } else if (root.right != null && root.left == null) {
            Result right = recurse(root.right);
            result.maxPath = Math.max(root.val, root.val + right.maxPath);
            result.maxSum = Math.max(result.maxPath, right.maxSum);
        } else {
            Result left = recurse(root.left);
            Result right = recurse(root.right);
            result.maxPath = Math.max(root.val, Math.max(root.val + left.maxPath, root.val + right.maxPath));
            result.maxSum = Math.max(Math.max(result.maxPath, left.maxSum), Math.max(right.maxSum, left.maxPath + root.val + right.maxPath));
        }
        
        return result;
    }
    
    class Result {
        public int maxSum;
        public int maxPath;
        
        public Result(int maxSum, int maxPath) {
            this.maxPath = maxPath;
            this.maxSum = maxSum;
        }
    }
}
